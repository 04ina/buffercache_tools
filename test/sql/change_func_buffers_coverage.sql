--
-- Preparing
--
SET allow_in_place_tablespaces = true;
CREATE TABLESPACE test_tablespace LOCATION '';
RESET allow_in_place_tablespaces;

CREATE DATABASE test_database_1;
CREATE DATABASE test_database_2;

\c test_database_1 \\
CREATE EXTENSION buffercache_tools;

CREATE TABLE test_table_1(col integer) 
    WITH (autovacuum_enabled = off)
    TABLESPACE pg_default;
INSERT INTO test_table_1 
    SELECT 1 FROM generate_series(1,1000); 

CREATE VIEW tt_1 AS 
    SELECT blocknum, dirty, pinning, fork 
        FROM pg_show_relation_buffers('test_table_1') 
        ORDER BY fork, blocknum; 

CREATE TABLE test_table_2(col integer) 
    WITH (autovacuum_enabled = off)
    TABLESPACE test_tablespace;
INSERT INTO test_table_2 
    SELECT 1 FROM generate_series(1,1000); 

CREATE VIEW tt_2 AS 
    SELECT blocknum, dirty, pinning, fork 
        FROM pg_show_relation_buffers('test_table_2') 
        ORDER BY fork, blocknum; 

\c test_database_2 \\
CREATE EXTENSION buffercache_tools;

CREATE TABLE test_table_3(col integer) 
    WITH (autovacuum_enabled = off)
    TABLESPACE pg_default;
INSERT INTO test_table_3 
    SELECT 1 FROM generate_series(1,1000); 

CREATE VIEW tt_3 AS 
    SELECT blocknum, dirty, pinning, fork 
        FROM pg_show_relation_buffers('test_table_3') 
        ORDER BY fork, blocknum; 
VACUUM;
CHECKPOINT;

\c test_database_1 \\

VACUUM;
CHECKPOINT;

--
-- Tests
--

-- 
-- Check pg_change_buffer() buffers coverage
--
SELECT * FROM tt_1;
SELECT pg_change_buffer('mark_dirty',
    ( 
        SELECT buffernum 
            FROM pg_show_relation_buffers('test_table_1') 
            ORDER BY fork, blocknum
            LIMIT 1
    )::integer
);
SELECT * FROM tt_1;
SELECT pg_change_buffer('flush',
    ( 
        SELECT buffernum 
            FROM pg_show_relation_buffers('test_table_1') 
            ORDER BY fork, blocknum 
            LIMIT 1
    )::integer
);
SELECT * FROM tt_1;

-- 
-- Check pg_change_relation_fork_buffers() buffers coverage
--
SELECT * FROM tt_1;
SELECT pg_change_relation_fork_buffers('mark_dirty', 'test_table_1', 'main');
SELECT * FROM tt_1;
SELECT pg_change_relation_fork_buffers('flush', 'test_table_1', 'main');
SELECT * FROM tt_1;

-- 
-- Check pg_change_relation_buffers() buffers coverage
--
SELECT * FROM tt_1;
SELECT pg_change_relation_buffers('mark_dirty', 'test_table_1');
SELECT * FROM tt_1;
SELECT pg_change_relation_buffers('flush', 'test_table_1');
SELECT * FROM tt_1;

-- 
-- Check pg_change_database_buffers() buffers coverage 
--
SELECT * FROM tt_1;
SELECT * FROM tt_2;
\c test_database_2 \\
SELECT * FROM tt_3;
\c test_database_1 \\

SELECT pg_change_database_buffers('mark_dirty',
    (
        SELECT oid 
            FROM pg_database 
            WHERE datname = 'test_database_1'
    )::oid  
);

SELECT * FROM tt_1;
SELECT * FROM tt_2;
\c test_database_2 \\
SELECT * FROM tt_3;
\c test_database_1 \\

SELECT pg_change_database_buffers('flush',
    (
        SELECT oid 
            FROM pg_database 
            WHERE datname = 'test_database_1'
    )::oid  
);

SELECT * FROM tt_1;
SELECT * FROM tt_2;
\c test_database_2 \\
SELECT * FROM tt_3;
\c test_database_1 \\

-- 
-- Check pg_change_tablespace_buffers() buffers coverage 
--
SELECT * FROM tt_1;
SELECT * FROM tt_2;
\c test_database_2 \\
SELECT * FROM tt_3;
\c test_database_1 \\

SELECT pg_change_tablespace_buffers('mark_dirty',
    (
        SELECT oid 
            FROM pg_tablespace 
            WHERE spcname = 'test_tablespace'
    )::oid  
);

SELECT * FROM tt_1;
SELECT * FROM tt_2;
\c test_database_2 \\
SELECT * FROM tt_3;
\c test_database_1 \\

SELECT pg_change_tablespace_buffers('flush',
    (
        SELECT oid 
            FROM pg_tablespace 
            WHERE spcname = 'test_tablespace'
    )::oid  
);

SELECT * FROM tt_1;
SELECT * FROM tt_2;
\c test_database_2 \\
SELECT * FROM tt_3;
\c test_database_1 \\

-- 
-- Check pg_change_all_valid_buffers() buffers coverage 
--
SELECT * FROM tt_1;
SELECT * FROM tt_2;
\c test_database_2 \\
SELECT * FROM tt_3;
\c test_database_1 \\

SELECT pg_change_all_valid_buffers('mark_dirty');

SELECT * FROM tt_1;
SELECT * FROM tt_2;
\c test_database_2 \\
SELECT * FROM tt_3;
\c test_database_1 \\

SELECT pg_change_all_valid_buffers('flush');

SELECT * FROM tt_1;
SELECT * FROM tt_2;
\c test_database_2 \\
SELECT * FROM tt_3;
\c test_database_1 \\

-- 
-- Check pg_change_buffer_by_page() buffers coverage
--
SELECT * FROM tt_1;
SELECT pg_change_buffer_by_page('mark_dirty', 'test_table_1', 'main', '0');
SELECT * FROM tt_1;
SELECT pg_change_buffer_by_page('flush', 'test_table_1', 'main', '0');
SELECT * FROM tt_1;

--
-- Cleanup 
--
DROP TABLE test_table_1;
DROP TABLE test_table_2;
\c test_database_2 \\
DROP TABLE test_table_3;

DROP TABLESPACE test_tablespace;
\c template1 \\

DROP DATABASE test_database_1;
DROP DATABASE test_database_2;