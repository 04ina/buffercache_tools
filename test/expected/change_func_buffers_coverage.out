--
-- Preparing
--
SET allow_in_place_tablespaces = true;
CREATE TABLESPACE test_tablespace LOCATION '';
RESET allow_in_place_tablespaces;
CREATE DATABASE test_database_1;
CREATE DATABASE test_database_2;
\c test_database_1 \\
CREATE EXTENSION buffercache_tools;
CREATE TABLE test_table_1(col integer) 
    WITH (autovacuum_enabled = off)
    TABLESPACE pg_default;
INSERT INTO test_table_1 
    SELECT 1 FROM generate_series(1,1000); 
CREATE VIEW tt_1 AS 
    SELECT blocknum, dirty, pinning, fork 
        FROM pg_show_relation_buffers('test_table_1') 
        ORDER BY fork, blocknum; 
CREATE TABLE test_table_2(col integer) 
    WITH (autovacuum_enabled = off)
    TABLESPACE test_tablespace;
INSERT INTO test_table_2 
    SELECT 1 FROM generate_series(1,1000); 
CREATE VIEW tt_2 AS 
    SELECT blocknum, dirty, pinning, fork 
        FROM pg_show_relation_buffers('test_table_2') 
        ORDER BY fork, blocknum; 
\c test_database_2 \\
CREATE EXTENSION buffercache_tools;
CREATE TABLE test_table_3(col integer) 
    WITH (autovacuum_enabled = off)
    TABLESPACE pg_default;
INSERT INTO test_table_3 
    SELECT 1 FROM generate_series(1,1000); 
CREATE VIEW tt_3 AS 
    SELECT blocknum, dirty, pinning, fork 
        FROM pg_show_relation_buffers('test_table_3') 
        ORDER BY fork, blocknum; 
VACUUM;
CHECKPOINT;
\c test_database_1 \\
VACUUM;
CHECKPOINT;
--
-- Tests
--
-- 
-- Check pg_change_buffer() buffers coverage
--
SELECT * FROM tt_1;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

SELECT pg_change_buffer('mark_dirty',
    ( 
        SELECT buffernum 
            FROM pg_show_relation_buffers('test_table_1') 
            ORDER BY fork, blocknum
            LIMIT 1
    )::integer
);
 pg_change_buffer 
------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | t     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

SELECT pg_change_buffer('flush',
    ( 
        SELECT buffernum 
            FROM pg_show_relation_buffers('test_table_1') 
            ORDER BY fork, blocknum 
            LIMIT 1
    )::integer
);
 pg_change_buffer 
------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

-- 
-- Check pg_change_relation_fork_buffers() buffers coverage
--
SELECT * FROM tt_1;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

SELECT pg_change_relation_fork_buffers('mark_dirty', 'test_table_1', 'main');
 pg_change_relation_fork_buffers 
---------------------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | t     |       0 | main
        1 | t     |       0 | main
        2 | t     |       0 | main
        3 | t     |       0 | main
        4 | t     |       0 | main
        0 | f     |       0 | vm
(9 rows)

SELECT pg_change_relation_fork_buffers('flush', 'test_table_1', 'main');
 pg_change_relation_fork_buffers 
---------------------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

-- 
-- Check pg_change_relation_buffers() buffers coverage
--
SELECT * FROM tt_1;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

SELECT pg_change_relation_buffers('mark_dirty', 'test_table_1');
 pg_change_relation_buffers 
----------------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | t     |       0 | fsm
        1 | t     |       0 | fsm
        2 | t     |       0 | fsm
        0 | t     |       0 | main
        1 | t     |       0 | main
        2 | t     |       0 | main
        3 | t     |       0 | main
        4 | t     |       0 | main
        0 | t     |       0 | vm
(9 rows)

SELECT pg_change_relation_buffers('flush', 'test_table_1');
 pg_change_relation_buffers 
----------------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

-- 
-- Check pg_change_database_buffers() buffers coverage 
--
SELECT * FROM tt_1;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

SELECT * FROM tt_2;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

\c test_database_2 \\
SELECT * FROM tt_3;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

\c test_database_1 \\
SELECT pg_change_database_buffers('mark_dirty',
    (
        SELECT oid 
            FROM pg_database 
            WHERE datname = 'test_database_1'
    )::oid  
);
 pg_change_database_buffers 
----------------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | t     |       0 | fsm
        1 | t     |       0 | fsm
        2 | t     |       0 | fsm
        0 | t     |       0 | main
        1 | t     |       0 | main
        2 | t     |       0 | main
        3 | t     |       0 | main
        4 | t     |       0 | main
        0 | t     |       0 | vm
(9 rows)

SELECT * FROM tt_2;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | t     |       0 | fsm
        1 | t     |       0 | fsm
        2 | t     |       0 | fsm
        0 | t     |       0 | main
        1 | t     |       0 | main
        2 | t     |       0 | main
        3 | t     |       0 | main
        4 | t     |       0 | main
        0 | t     |       0 | vm
(9 rows)

\c test_database_2 \\
SELECT * FROM tt_3;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

\c test_database_1 \\
SELECT pg_change_database_buffers('flush',
    (
        SELECT oid 
            FROM pg_database 
            WHERE datname = 'test_database_1'
    )::oid  
);
 pg_change_database_buffers 
----------------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

SELECT * FROM tt_2;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

\c test_database_2 \\
SELECT * FROM tt_3;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

\c test_database_1 \\
-- 
-- Check pg_change_tablespace_buffers() buffers coverage 
--
SELECT * FROM tt_1;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

SELECT * FROM tt_2;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

\c test_database_2 \\
SELECT * FROM tt_3;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

\c test_database_1 \\
SELECT pg_change_tablespace_buffers('mark_dirty',
    (
        SELECT oid 
            FROM pg_tablespace 
            WHERE spcname = 'test_tablespace'
    )::oid  
);
 pg_change_tablespace_buffers 
------------------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

SELECT * FROM tt_2;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | t     |       0 | fsm
        1 | t     |       0 | fsm
        2 | t     |       0 | fsm
        0 | t     |       0 | main
        1 | t     |       0 | main
        2 | t     |       0 | main
        3 | t     |       0 | main
        4 | t     |       0 | main
        0 | t     |       0 | vm
(9 rows)

\c test_database_2 \\
SELECT * FROM tt_3;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

\c test_database_1 \\
SELECT pg_change_tablespace_buffers('flush',
    (
        SELECT oid 
            FROM pg_tablespace 
            WHERE spcname = 'test_tablespace'
    )::oid  
);
 pg_change_tablespace_buffers 
------------------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

SELECT * FROM tt_2;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

\c test_database_2 \\
SELECT * FROM tt_3;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

\c test_database_1 \\
-- 
-- Check pg_change_all_valid_buffers() buffers coverage 
--
SELECT * FROM tt_1;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

SELECT * FROM tt_2;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

\c test_database_2 \\
SELECT * FROM tt_3;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

\c test_database_1 \\
SELECT pg_change_all_valid_buffers('mark_dirty');
 pg_change_all_valid_buffers 
-----------------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | t     |       0 | fsm
        1 | t     |       0 | fsm
        2 | t     |       0 | fsm
        0 | t     |       0 | main
        1 | t     |       0 | main
        2 | t     |       0 | main
        3 | t     |       0 | main
        4 | t     |       0 | main
        0 | t     |       0 | vm
(9 rows)

SELECT * FROM tt_2;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | t     |       0 | fsm
        1 | t     |       0 | fsm
        2 | t     |       0 | fsm
        0 | t     |       0 | main
        1 | t     |       0 | main
        2 | t     |       0 | main
        3 | t     |       0 | main
        4 | t     |       0 | main
        0 | t     |       0 | vm
(9 rows)

\c test_database_2 \\
SELECT * FROM tt_3;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | t     |       0 | fsm
        1 | t     |       0 | fsm
        2 | t     |       0 | fsm
        0 | t     |       0 | main
        1 | t     |       0 | main
        2 | t     |       0 | main
        3 | t     |       0 | main
        4 | t     |       0 | main
        0 | t     |       0 | vm
(9 rows)

\c test_database_1 \\
SELECT pg_change_all_valid_buffers('flush');
 pg_change_all_valid_buffers 
-----------------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

SELECT * FROM tt_2;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

\c test_database_2 \\
SELECT * FROM tt_3;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

\c test_database_1 \\
-- 
-- Check pg_change_buffer_by_page() buffers coverage
--
SELECT * FROM tt_1;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

SELECT pg_change_buffer_by_page('mark_dirty', 'test_table_1', 'main', '0');
 pg_change_buffer_by_page 
--------------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | t     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

SELECT pg_change_buffer_by_page('flush', 'test_table_1', 'main', '0');
 pg_change_buffer_by_page 
--------------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknum | dirty | pinning | fork 
----------+-------+---------+------
        0 | f     |       0 | fsm
        1 | f     |       0 | fsm
        2 | f     |       0 | fsm
        0 | f     |       0 | main
        1 | f     |       0 | main
        2 | f     |       0 | main
        3 | f     |       0 | main
        4 | f     |       0 | main
        0 | f     |       0 | vm
(9 rows)

--
-- Cleanup 
--
DROP TABLE test_table_1;
DROP TABLE test_table_2;
\c test_database_2 \\
DROP TABLE test_table_3;
DROP TABLESPACE test_tablespace;
\c template1 \\
DROP DATABASE test_database_1;
DROP DATABASE test_database_2;
